packages/newsroom-signup/src/DataWrapper.tsx

```diff
@@ -0,0 +1,108 @@
+  query {
+    currentUser {
+      ethAddress
+    }
+  }
+`;
+// Unfortunately we can't combine this charter query with eth address query, because if no charter is saved at all yet, then the whole query errors and we won't get eth address
+const getCharterQuery = gql`
+  query {
+    jsonb(id: "${JSON_BLOB_ID}") {
+      rawJson
+    }
+  }
+`;
+
+const saveCharterMutation = gql`
+  mutation($input: JsonbInput!) {
+    jsonbSave(input: $input) {
+      id
+    }
+  }
+`;
+
+export interface DataWrapperChildrenProps {
+  profileWalletAddress: EthAddress;
+  persistedCharter: Partial<CharterData>;
+  persistCharter(charter: Partial<CharterData>): Promise<void>;
+}
+
+export interface DataWrapperProps {
+  children(props: DataWrapperChildrenProps): any;
+}
+
+export class DataWrapper extends React.Component<DataWrapperProps> {
+  public render(): JSX.Element {
+    return (
+      <Query query={userEthAddress}>
+        {({ loading, error, data }) => {
+          if (loading) {
+            return "Loading...";
+          }
+          if (error) {
+            return `Error! ${JSON.stringify(error)}`;
+          }
+
+          return (
+            <Query query={getCharterQuery}>
+              {({ loading: charterLoading, error: charterError, data: charterData }) => {
+                if (charterLoading) {
+                  return "Loading...";
+                }
+                if (charterError) {
+                  if (charterError.graphQLErrors && charterError.graphQLErrors[0].message === "No jsonb found") {
+                    // ok, they haven't saved a charter yet
+                  } else {
+                    return `Error! ${JSON.stringify(charterError)}`;
+                  }
+                }
+
+                let persistedCharter: Partial<CharterData>;
+                if (charterData && charterData.jsonb && charterData.jsonb[0].rawJson) {
+                  try {
+                    persistedCharter = JSON.parse(charterData.jsonb[0].rawJson);
+                  } catch (err) {
+                    console.error("Failed to parse persisted charter JSON", err, charterData.jsonb.rawJson);
+                  }
+                }
+
+                return (
+                  <Mutation mutation={saveCharterMutation}>
+                    {(saveCharter: MutationFunc) => {
+                      return this.props.children({
+                        profileWalletAddress: data.currentUser.ethAddress,
+                        persistedCharter,
+                        persistCharter: this.saveCharterFuncFromMutation(saveCharter),
+                      });
+                    }}
+                  </Mutation>
+                );
+              }}
+            </Query>
+          );
+        }}
+      </Query>
+    );
+  }
+
+  private saveCharterFuncFromMutation(mutation: MutationFunc): (charter: Partial<CharterData>) => Promise<void> {
+    return async (charter: Partial<CharterData>) => {
+      await mutation({
```

Comment:
shouldn't this return?

Reply:
Yep good call. Newsroom component isn't using return value of `persistCharter` at the moment but makes sense to pass it through anyway.